In [64]:
import pandas as pd 

df = pd.read_csv('business.csv')
df.head()

,date,airline,ch_code,num_code,dep_time,from,time_taken,stop,arr_time,to,price
0,11-02-2022,Air India,AI,868,18:00,Delhi,02h 00m,non-stop,20:00,Mumbai,"25,612"
1,11-02-2022,Air India,AI,624,19:00,Delhi,02h 15m,non-stop,21:15,Mumbai,"25,612"
2,11-02-2022,Air India,AI,531,20:00,Delhi,24h 45m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,20:45,Mumbai,"42,220"
3,11-02-2022,Air India,AI,839,21:25,Delhi,26h 30m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,23:55,Mumbai,"44,450"
4,11-02-2022,Air India,AI,544,17:15,Delhi,06h 40m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,23:55,Mumbai,"46,690"


In [65]:
## we are here and main aim is to price prediction and so we are focus on price column

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 93487 entries, 0 to 93486
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   date        93487 non-null  str  
 1   airline     93487 non-null  str  
 2   ch_code     93487 non-null  str  
 3   num_code    93487 non-null  int64
 4   dep_time    93487 non-null  str  
 5   from        93487 non-null  str  
 6   time_taken  93487 non-null  str  
 7   stop        93487 non-null  str  
 8   arr_time    93487 non-null  str  
 9   to          93487 non-null  str  
 10  price       93487 non-null  str  
dtypes: int64(1), str(10)
memory usage: 7.8 MB


In [66]:
df.describe()

,num_code
count,93487.000000
mean,780.056147
std,147.616038
min,401.000000
25%,706.000000
50%,820.000000
75%,874.000000
max,996.000000


## 3️⃣ Date, Month, Year ko Separate Kyu Aur Kaise Karte Hain? (Feature Engineering)

### ❓ Kyu alag karte hain?
Machine Learning model ko agar aap direct `'11-02-2022'` doge, to model ko ye sirf ek bekaar text lagega, wo isse kuch seekh nahi payega.

Lekin agar aap ise todkar alag-alag columns bana doge:
* **`Date` (Day):** `11`
* **`Month`:** `2` (February)
* **`Year`:** `2022`

To Model aasani se pattern pakad lega (jaise: *Festivals/Holidays wale month me ticket mehengi hoti hai, ya weekend wali date par price badh jata hai*).

---

### 🛠️ Hume Kaise Karna Hoga? (Code)

Aapke `df['date']` me format hai: `'11-02-2022'` (`DD-MM-YYYY`).

Ise alag karne ke **2 simple tareeqe** hain:

#### 🔹 Tarika 1: String Split Method (`.str.split`)
```python
# 1. Date (Day) nikaalna
df['Date'] = df['date'].str.split('-').str[0].astype(int)

# 2. Month nikaalna
df['Month'] = df['date'].str.split('-').str[1].astype(int)

# 3. Year nikaalna
df['Year'] = df['date'].str.split('-').str[2].astype(int)

# 4. Ab purane 'date' text column ko delete kar do
df.drop('date', axis=1, inplace=True)

df[['Date', 'Month', 'Year']].head()
```

#### 🔹 Tarika 2: Pandas DateTime Method (Professional Way)
```python
# 'date' column ko datetime me convert karein
df['date'] = pd.to_datetime(df['date'], format='%d-%m-%Y')

# Date, Month, Year extract karein
df['Date'] = df['date'].dt.day
df['Month'] = df['date'].dt.month
df['Year'] = df['date'].dt.year

# Original date column drop karein
df.drop('date', axis=1, inplace=True)
```

---

### 💡 Ek Important Tip aapke dataset ke liye (`price` column):
Aapke `df['price']` me comma laga hua hai (jaise `'25,612'`), isliye wo `str` ban gaya hai. Use number banane ke liye yeh zaroor karna:

```python
# Comma hatakar Integer/Float me convert karein
df['price'] = df['price'].str.replace(',', '').astype(int)
```

In [67]:
#feature engineering 

# 1. Date (Day) nikaalna
df['Date'] = df['date'].str.split('-').str[0].astype(int)
# 2. Month nikaalna
df['Month'] = df['date'].str.split('-').str[1].astype(int)
# 3. Year nikaalna
df['Year'] = df['date'].str.split('-').str[2].astype(int)
# 4. Ab purane 'date' text column ko delete kar do
df.drop('date', axis=1, inplace=True)
df[['Date', 'Month', 'Year']].head()


,Date,Month,Year
0,11,2,2022
1,11,2,2022
2,11,2,2022
3,11,2,2022
4,11,2,2022


In [68]:
df.head()

,airline,ch_code,num_code,dep_time,from,time_taken,stop,arr_time,to,price,Date,Month,Year
0,Air India,AI,868,18:00,Delhi,02h 00m,non-stop,20:00,Mumbai,"25,612",11,2,2022
1,Air India,AI,624,19:00,Delhi,02h 15m,non-stop,21:15,Mumbai,"25,612",11,2,2022
2,Air India,AI,531,20:00,Delhi,24h 45m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,20:45,Mumbai,"42,220",11,2,2022
3,Air India,AI,839,21:25,Delhi,26h 30m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,23:55,Mumbai,"44,450",11,2,2022
4,Air India,AI,544,17:15,Delhi,06h 40m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,23:55,Mumbai,"46,690",11,2,2022


In [69]:
df['price'] = df['price'].str.replace(',', '').astype(int)

In [70]:
df['arrival_hour'] = df['arr_time'].str.split(':').str[0].astype(int)
df['arrival_min'] = df['arr_time'].str.split(':').str[1].astype(int)

In [71]:
df.head()

,airline,ch_code,num_code,dep_time,from,time_taken,stop,arr_time,to,price,Date,Month,Year,arrival_hour,arrival_min
0,Air India,AI,868,18:00,Delhi,02h 00m,non-stop,20:00,Mumbai,25612,11,2,2022,20,0
1,Air India,AI,624,19:00,Delhi,02h 15m,non-stop,21:15,Mumbai,25612,11,2,2022,21,15
2,Air India,AI,531,20:00,Delhi,24h 45m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,20:45,Mumbai,42220,11,2,2022,20,45
3,Air India,AI,839,21:25,Delhi,26h 30m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,23:55,Mumbai,44450,11,2,2022,23,55
4,Air India,AI,544,17:15,Delhi,06h 40m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,23:55,Mumbai,46690,11,2,2022,23,55


In [72]:
df.drop('arr_time', axis=1, inplace=True)

In [73]:
df.head()

,airline,ch_code,num_code,dep_time,from,time_taken,stop,to,price,Date,Month,Year,arrival_hour,arrival_min
0,Air India,AI,868,18:00,Delhi,02h 00m,non-stop,Mumbai,25612,11,2,2022,20,0
1,Air India,AI,624,19:00,Delhi,02h 15m,non-stop,Mumbai,25612,11,2,2022,21,15
2,Air India,AI,531,20:00,Delhi,24h 45m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,42220,11,2,2022,20,45
3,Air India,AI,839,21:25,Delhi,26h 30m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,44450,11,2,2022,23,55
4,Air India,AI,544,17:15,Delhi,06h 40m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,46690,11,2,2022,23,55


In [74]:
# now split the dep_time also into two parts hours, min

df['dep_hour'] = df['dep_time'].str.split(':').str[0].astype(int)
df['dep_min'] = df['dep_time'].str.split(':').str[1].astype(int)

In [75]:
df.head()

,airline,ch_code,num_code,dep_time,from,time_taken,stop,to,price,Date,Month,Year,arrival_hour,arrival_min,dep_hour,dep_min
0,Air India,AI,868,18:00,Delhi,02h 00m,non-stop,Mumbai,25612,11,2,2022,20,0,18,0
1,Air India,AI,624,19:00,Delhi,02h 15m,non-stop,Mumbai,25612,11,2,2022,21,15,19,0
2,Air India,AI,531,20:00,Delhi,24h 45m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,42220,11,2,2022,20,45,20,0
3,Air India,AI,839,21:25,Delhi,26h 30m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,44450,11,2,2022,23,55,21,25
4,Air India,AI,544,17:15,Delhi,06h 40m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,46690,11,2,2022,23,55,17,15


In [76]:
df.drop(columns=['dep_time'], inplace=True)


In [77]:
df.head()

,airline,ch_code,num_code,from,time_taken,stop,to,price,Date,Month,Year,arrival_hour,arrival_min,dep_hour,dep_min
0,Air India,AI,868,Delhi,02h 00m,non-stop,Mumbai,25612,11,2,2022,20,0,18,0
1,Air India,AI,624,Delhi,02h 15m,non-stop,Mumbai,25612,11,2,2022,21,15,19,0
2,Air India,AI,531,Delhi,24h 45m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,42220,11,2,2022,20,45,20,0
3,Air India,AI,839,Delhi,26h 30m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,44450,11,2,2022,23,55,21,25
4,Air India,AI,544,Delhi,06h 40m,1-stop\n\t\t\t\t\t\t\t\t\t\t\t\t\n\t\t\t\t\t\t...,Mumbai,46690,11,2,2022,23,55,17,15


In [78]:
df['stop'] = df['stop'].str.split().str[0]

df['stop'].unique()

array(['non-stop', '1-stop', '2+-stop'], dtype=object)

In [79]:
df['stop'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
93482    False
93483    False
93484    False
93485    False
93486    False
Name: stop, Length: 93487, dtype: bool

In [80]:
import numpy as np
df['Total_stops'] = df['stop'].map({'non-stop':0, '1-stop':1, '2+-stop':2, np.nan:1})

In [81]:
df[['stop', 'Total_stops']]

,stop,Total_stops
0,non-stop,0
1,non-stop,0
2,1-stop,1
3,1-stop,1
4,1-stop,1
...,...,...
93482,1-stop,1
93483,1-stop,1
93484,1-stop,1
93485,1-stop,1


In [82]:
df[df['Total_stops'].isnull()]

,airline,ch_code,num_code,from,time_taken,stop,to,price,Date,Month,Year,arrival_hour,arrival_min,dep_hour,dep_min,Total_stops


In [83]:
# wants the duration in hours or minutes totally

df['duration_hrs']= df['time_taken'].str.split(' ').str[0].str.replace('h','').astype(int)

# extract the minutes section
df['duration_mins'] = df['time_taken'].str.split(' ').str[1].str.replace('m', '').astype(int)

# total minutes calculating 
df['total_duration_in_mins'] = (df['duration_hrs'] * 60) + df['duration_mins']

# 4. Ab purane text wale 'time_taken' column ko drop kar dein
df.drop('time_taken', axis=1, inplace=True)


In [84]:
# 5. Check karein
df[['duration_hrs', 'duration_mins', 'total_duration_in_mins']].head()


,duration_hrs,duration_mins,total_duration_in_mins
0,2,0,120
1,2,15,135
2,24,45,1485
3,26,30,1590
4,6,40,400


## Now we have different different categories in each columns like airlines are IndiGo, AirIndia etc adn sources are Banglore, Delhi etc...

### so now for all these we use onehotencoder to make all these in numerical form

In [85]:
from sklearn.preprocessing import OneHotEncoder

In [86]:
encoder = OneHotEncoder()

In [87]:
df

,airline,ch_code,num_code,from,stop,to,price,Date,Month,Year,arrival_hour,arrival_min,dep_hour,dep_min,Total_stops,duration_hrs,duration_mins,total_duration_in_mins
0,Air India,AI,868,Delhi,non-stop,Mumbai,25612,11,2,2022,20,0,18,0,0,2,0,120
1,Air India,AI,624,Delhi,non-stop,Mumbai,25612,11,2,2022,21,15,19,0,0,2,15,135
2,Air India,AI,531,Delhi,1-stop,Mumbai,42220,11,2,2022,20,45,20,0,1,24,45,1485
3,Air India,AI,839,Delhi,1-stop,Mumbai,44450,11,2,2022,23,55,21,25,1,26,30,1590
4,Air India,AI,544,Delhi,1-stop,Mumbai,46690,11,2,2022,23,55,17,15,1,6,40,400
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
93482,Vistara,UK,822,Chennai,1-stop,Hyderabad,69265,31,3,2022,19,50,9,45,1,10,5,605
93483,Vistara,UK,826,Chennai,1-stop,Hyderabad,77105,31,3,2022,22,55,12,30,1,10,25,625
93484,Vistara,UK,832,Chennai,1-stop,Hyderabad,79099,31,3,2022,20,55,7,5,1,13,50,830
93485,Vistara,UK,828,Chennai,1-stop,Hyderabad,81585,31,3,2022,17,0,7,0,1,10,0,600


In [88]:
encoded_arr = encoder.fit_transform(df[['airline', 'from', 'to']]).toarray()

In [89]:
encoded_df = pd.DataFrame(encoded_arr, columns=encoder.get_feature_names_out())
encoded_df.head()


,airline_Air India,airline_Vistara,from_Bangalore,from_Chennai,from_Delhi,from_Hyderabad,from_Kolkata,from_Mumbai,to_Bangalore,to_Chennai,to_Delhi,to_Hyderabad,to_Kolkata,to_Mumbai
0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
1,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
2,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
3,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
4,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
